crypto/asn1_dsa.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/*
+ * A simple ASN.1 DER encoder/decoder for DSA-Sig-Value and ECDSA-Sig-Value.
+ *
+ * DSA-Sig-Value ::= SEQUENCE {
+ *  r  INTEGER,
+ *  s  INTEGER
+ * }
+ *
+ * ECDSA-Sig-Value ::= SEQUENCE {
+ *  r  INTEGER,
+ *  s  INTEGER
+ * }
+ */
+
+#include <openssl/bn.h>
+#include "internal/asn1_dsa.h"
+
+#define ID_SEQUENCE 0x30
+#define ID_INTEGER 0x02
+
+static size_t calc_der_len(size_t cont_len)
+{
+    size_t llen;
+
+    if (cont_len <= 0x7f)
+        llen = 1;
+    else if (cont_len <= 0xff)
+        llen = 2;
+    else if (cont_len <= 0xffff)
+        llen = 3;
+    else
+        return 0;
+    return 1 + llen + cont_len;
+}
+
+static void enc_len(size_t cont_len, unsigned char **ppout)
+{
+    if (cont_len > 0xff) {
+        *(*ppout)++ = 0x82;
+        *(*ppout)++ = (unsigned char)(cont_len >> 8);
+    } else if (cont_len > 0x7f)
+        *(*ppout)++ = 0x81;
+    *(*ppout)++ = (unsigned char)cont_len;
+}
+
+static size_t calc_int_der_len(const BIGNUM *n)
+{
+    if (BN_is_negative(n))
+        return 0;
+    return calc_der_len(BN_num_bits(n) / 8 + 1);
```

Comment:
Hint: there is a `BN_num_bytes()`

Reply:
There is a difference. `BN_num_bytes(a)` is `(BN_num_bits(a)+7)/8`; it rounds up. I wanted to round down and add one in order to add a padding byte *iff* `BN_num_bits()` is a multiple of 8 to ensure valid INTEGER encoding for positive values (`r` and `s` are always positive).

Reply:
Ok, makes some kind of sense...

Reply:
Perhaps add a comment to avoid future confusion.

Reply:
I have created a macro to calculate the length and added a descriptive comment.